### PR TITLE
Add ownerReferences to virt-launcher

### PR DESCRIPTION
**Pull request by shiywang**
Signed-off-by: Shiyang Wang <shiywang@redhat.com>

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubevirt/kubevirt/issues/1369

**Special notes for your reviewer**:

**Release note**:

```release-note
add ownerReferences field to virt-launcher
```

### CODE REVIEW

**Comment by cynepco3hahue** [comment]
@shiywang I think it already fixed by https://github.com/kubevirt/kubevirt/pull/1320.

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -378,6 +378,7 @@ func (t *templateService) RenderLaunchManifest(vmi *v1.VirtualMachineInstance) (
 	hostName := dns.SanitizeHostname(vmi)
 
 	// TODO use constants for podLabels
+	ts := true
```
Comment (fabiand) [inline pkg/virt-controller/services/template.go]: Why is it named `ts`? Why did you use a variable for this?

**Comment by fabiand** [comment]
+1 on David's comment.

Just adding the ownerReference is not enough, we want to deprecate the alternative mechanism we have in place right now.

**Comment by shiywang** [comment]
@fabiand @davidvossel I updated a little bit, could you please give a quick glance to see if I'm changing the right way, then I'll add the unit/functional tests and make the CI pass.

**Comment by davidvossel** [comment]
@rmohr Was there a reason we chose not to use owner references for the VMI's pod? Did we have some issue with cascading deletion or something?

**Comment by rmohr** [comment]
> @rmohr Was there a reason we chose not to use owner references for the VMI's pod? Did we have some issue with cascading deletion or something?

@davidvossel When I last rewrote the controller I was thinking about adding it. I decided not to back then because I did not want to make the PR more complex. The main reason was that "orphan delete" and "background delete" are not really something which properly fit the relationship between a Pod and a VMI (orphan delete of a VMI is basically a force delete of a pod).

However, now we have a finalizer on the VMI. No matter how the garbage collector will try to delete the VMI, we will have the chance to do final cleanup (making sure that the pod goes down too). So we should be save now to use owner references.

File: pkg/virt-controller/watch/vmi.go [hunk 2]
```diff
@@ -573,13 +584,35 @@ func (c *VMIController) filterMatchingPods(vmi *virtv1.VirtualMachineInstance, p
...
+
 func isPodOwnedByHandler(pod *k8sv1.Pod) bool {
-	if pod.Annotations != nil && pod.Annotations[virtv1.OwnedByAnnotation] == "virt-handler" {
+	if pod.OwnerReferences != nil && pod.OwnerReferences[0].Kind == "Pod" && strings.Contains(pod.OwnerReferences[0].Name, "virt-handler") {
 		return true
 	}
 	return false
 }
 
+func updatePodOwnerToHandlerPod(pod, handlerPod *k8sv1.Pod) {
```
Comment (rmohr) [inline pkg/virt-controller/watch/vmi.go]: Hm, that approach looks in this context wrong. `virt-handler` is then responsible, that's true but still the VMI object is the owner of the pod. I am not sure if we can replace the annotation.

**Comment by shiywang** [comment]
>The main reason was that "orphan delete" and "background delete" are not really something which properly fit the relationship between a Pod and a VMI (orphan delete of a VMI is basically a force delete of a pod).

Agree, so no matter what flags you use --cascade=false/true it should be act as --cascade=true, and `propagationPolicy` always foreground deletion. (in other words, no orphan, no background delete) right ?

>However, now we have a finalizer on the VMI. No matter how the garbage collector will try to delete the VMI, we will have the chance to do final cleanup (making sure that the pod goes down too). So we should be save now to use owner references.

So no matter what kind of cascade deletion mechanism,  what kind of propagation we use, code [here](https://github.com/kubevirt/kubevirt/blob/129e15349012aa7da11a367db57f8b29206bee74/pkg/virt-controller/watch/vmi.go#L394) will make sure pod being deleted and finalizer on VMI will make sure VMI resource being deleted after pod, right ?

**Comment by rmohr** [comment]
> So no matter what kind of cascade deletion mechanism, what kind of propagation we use, code here will make sure pod being deleted and finalizer on VMI will make sure VMI resource being deleted after pod, right ?

Exactly. :+1:

**Comment by rmohr** [comment]
I would be ok with taking this (that's  the most interesting part of the PR because it is the external visible part with a meaning for users/uis/...) and defer the annotation removal to a different PR.

A follow up PR or this one can do the following to get rid of the annotation:

 * If the VMI is in `scheduling` state and `isPodReady` are true, then move the VMI to scheduled
 * remove all traces of `isPodOwnedByHandler` and replace with `isPodReady` where appropriate (I think only necessary in `updateStatus`)
 * still keep the handOver expectator and let it act on the transition on the VMI from `scheduling` to `scheduled`

@shiywang what do you think?

**Comment by rmohr** [comment]
ci test please

**Comment by fabiand** [comment]
This PR has been stale for a hwile. I asked @ahadas to pick this up.

**Comment by rmohr** [comment]
Closing this in favour of #1641 which got merged. @shiywang thanks for starting it.

**Pull request closed without merging**